### PR TITLE
Add Label type

### PR DESCRIPTION
**Pull request by TheRawMeatball**
Maybe potentially solves #103 and some similar issues.
Using these for RenderGraph nodes and slots is planned.

### CODE REVIEW

**Comment by Ratysz** [comment]
This looks way more complex than it needs to be.

Why not have separate derivable traits for labels of different types, instead of requiring an extra argument for the derive macro? E.g.,
```rust
#[derive(Debug, Hash, PartialEq, Eq, Clone, IntoLabel)]
#[label_type(StageLabel)]
enum Stages {
    SceneStage,
}
```
could be just 
```rust
#[derive(Debug, Hash, PartialEq, Eq, Clone, StageLabel)]
enum Stages {
    SceneStage,
}
```
and `Label<StageLabel>` could then be reduced to
```rust
type BoxedStageLabel = Box<dyn StageLabel>;
```
(I'm not sure why does it have to be an `Arc` in the PR). Ghastly-named `trait FullIntoLabel` then disappears entirely, leaving only `label: impl StageLabel` or `label: impl SystemLabel` in it's place.

`StageLabel` and `SystemLabel` (and whichever else we might need) can share most of their machinery, you'll only need to implement `PartialEq`, `Eq`, and `Hash` on each `dyn` of them. `Clone` can also be arranged. See [here](https://github.com/Ratysz/bevy_prototype_scheduler/blob/stage_label_dyn_hacks/src/stage_label.rs) for reference.

To prevent getting stuck beyond the bikeshedding event horizon of infinite potential, I would leave replacing all labels out of the PR, and instead implemented `StageLabel` (and maybe `SystemLabel`) on `Into<Cow<'static, str>>`. This should allow leaving them as is for now, and change them out slowly as we need - chances are we'll ditch stages entirely before such need arises anyway.

Unrelated, system labels don't seem to be plumbed in everywhere.

**Comment by alice-i-cecile** [comment]
Being able to refer to other systems by their type in some form rather than relying on always having to explicitly label your systems would be a really impressive win from an ergonomics perspective.

This would make the beginner experience a lot nicer and avoid really tedious boilerplate.

You'll usually only have one system for each function, and dependencies will usually be defined between systems in the same module.

I'm not super qualified on how you'd *do* so (default the label field using reflection?) but feel free to bug #ecs if you need mad science help and I can dig in.

Otherwise, Ratys's comments look solid.

**Comment by Ratysz** [comment]
We don't have a better way of doing this. Type name, or anything derived from just the signature, is about the only unambiguous thing we can get from a function without `IntoSystem`-ing it.

In any case, this is insufficient, because we may need multiple instances of the same type: not everything that's a system is a distinct function. It also muddles encapsulation of systems' labels and encapsulation of systems themselves: a plugin might not want to expose a system, to prevent multiple instances of it (potentially a safety concern), but would like to allow its users to order their systems relative to it.

> This would make the beginner experience a lot nicer and avoid really tedious boilerplate.

I'll raise my point from Discord again: putting the schedule together is not something a user would be doing constantly. Barring pathological cases, any boilerplate we may introduce will stand out only in sterile demo code.

**Review by Ratysz — CHANGES_REQUESTED** [review]
I still think that changing all the labels everywhere should be a separate PR: it's a lot of small, bikesheddable changes everywhere, it distracts from the implementation itself.

Tests should definitely not be touched: we continue to enable the simple stringly-typed labels for a reason.

File: crates/bevy_ecs/src/schedule/label.rs [hunk 1]
```diff
@@ -13,7 +13,7 @@ pub trait Label<T>: DynHash + DynClone<T> + Send + Sync + 'static {
 
 impl<M: 'static> Debug for dyn Label<M> {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        f.write_str(&self.name())
+        f.write_str(&format!("\"{}\"", &self.name()))
```
Comment (bjorn3) [inline crates/bevy_ecs/src/schedule/label.rs:16]:
```suggestion
        write!(f, "{:?}", self.name())
```

This will also put it between `"`, but avoids an allocation and will escapes `"` and special characters.

File: crates/bevy_ecs/src/schedule/label.rs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    any::Any,
+    borrow::Cow,
+    fmt::Debug,
+    hash::{Hash, Hasher},
+};
+
+use crate::{StageLabelMarker, SystemLabelMarker};
+
+pub trait Label<T>: DynHash + DynClone<T> + Send + Sync + 'static {
+    fn name(&self) -> Cow<'static, str>;
```
[the remark this answers is not in the document]
Comment (TheRawMeatball) [reply crates/bevy_ecs/src/schedule/label.rs:11]: Using a special `name` allows us to return a `Cow<'static, str>` without, which avoids an allocation when we need to access the name

File: crates/bevy_ecs/src/schedule/stage.rs [hunk 3]
```diff
@@ -350,8 +350,8 @@ impl SystemStage {
 }
 
 enum DependencyGraphError {
-    LabelNotFound(Cow<'static, str>),
-    DuplicateLabel(Cow<'static, str>),
+    LabelNotFound(SystemLabel),
+    DuplicateLabel(SystemLabel),
     GraphCycles(Vec<Cow<'static, str>>),
```
[the remark this answers is not in the document]
Comment (TheRawMeatball) [reply crates/bevy_ecs/src/schedule/stage.rs:355]: `Cow<'static, str>` is needed so we can substitute the system name when there isn't a label

File: crates/bevy_ecs/src/schedule/system_descriptor.rs [hunk 4]
```diff
@@ -15,21 +16,28 @@ use std::borrow::Cow;
 /// that they have to run before or after the system with that label.
 ///
 /// # Example
-/// ```rust
+/// ```ignore
+/// # // This is no_run because the macro doesn't like the doctest environment
```
[the remark this answers is not in the document]
Comment (TheRawMeatball) [reply crates/bevy_ecs/src/schedule/system_descriptor.rs:20]: Hmm, maybe #1434 is related?

File: crates/bevy_ecs/src/schedule/stage.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ratysz) [reply crates/bevy_ecs/src/schedule/stage.rs:355]: Ah, right; this is correct.

**Comment by TheRawMeatball** [comment]
> I still think that changing all the labels everywhere should be a separate PR: it's a lot of small, bikesheddable changes everywhere, it distracts from the implementation itself.

They're changes that need to be done eventually, why wait? They're already done
> Tests should definitely not be touched: we continue to enable the simple stringly-typed labels for a reason.

That's a fair point though, I'll revert the changes in tests

File: crates/bevy_ecs/src/schedule/label.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ratysz) [reply crates/bevy_ecs/src/schedule/label.rs:11]: I agree, this is cleaner. However, it'd be amazing if there was some way to make the macro's default use `Debug`/`Display` somehow. (I'm not entirely sure what exactly does it do now, use the type name?)

File: crates/bevy_ecs/src/schedule/system_descriptor.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (TheRawMeatball) [reply crates/bevy_ecs/src/schedule/system_descriptor.rs:20]: Nevermind, looks unrelated

**Comment by Ratysz** [comment]
> They're changes that need to be done eventually, why wait? They're already done

Yes, I forgot to add "but they're already done, so never mind" - I don't think you should revert them now, but I haven't changed my opinion on them in general.

File: crates/bevy_ecs/src/schedule/label.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (TheRawMeatball) [reply crates/bevy_ecs/src/schedule/label.rs:11]: Currently, it uses the type name, but allows the user to inject an arbitrary expression with `#[name_expr]` so it can be overriden. I'm not sure if we want that API though, so its uncertain

File: crates/bevy_ecs/src/schedule/mod.rs [hunk 5]
```diff
@@ -127,32 +157,38 @@ impl Schedule {
...
         func: F,
     ) -> &mut Self {
+        let name = name.into();
         let stage = self
-            .get_stage_mut::<T>(name)
-            .unwrap_or_else(|| panic!("stage '{}' does not exist or is the wrong type", name));
+            .get_stage_mut::<T, _>(name.clone())
+            .unwrap_or_else(move || {
+                panic!(
+                    "stage '{}' does not exist or is the wrong type",
```
[the remark this answers is not in the document]
Comment (TheRawMeatball) [reply crates/bevy_ecs/src/schedule/mod.rs:168]: If it's out of scope, this TODO should probably be added separately, or filed as an issue

File: crates/bevy_ecs/src/schedule/mod.rs [hunk 6]
```diff
@@ -127,32 +157,38 @@ impl Schedule {
...
+                    name.name()
+                )
+            });
         func(stage);
         self
     }
 
-    pub fn get_stage<T: Stage>(&self, name: &str) -> Option<&T> {
+    pub fn get_stage<T: Stage>(&self, name: impl Into<StageLabel>) -> Option<&T> {
```
Comment (Ratysz) [inline crates/bevy_ecs/src/schedule/mod.rs:176]: This and `get_stage_mut()` have different generics in signatures.

File: crates/bevy_ecs/src/schedule/stage.rs [hunk 7]
```diff
@@ -503,17 +503,17 @@ fn find_ambiguities(systems: &[impl SystemContainer]) -> Vec<(usize, usize)> {
...
     for (index_a, relations) in all_relations.drain(..).enumerate() {
         // TODO: prove that `.take(index_a)` would be correct here, and uncomment it if so.
         for index_b in full_bitset.difference(&relations)
         /*.take(index_a)*/
         {
-            if !processed.contains(index_b) && !systems[index_a].is_compatible(&systems[index_b]) {
+            if !systems[index_a].is_compatible(&systems[index_b])
+                && !ambiguities.contains(&(index_b, index_a))
+            {
                 ambiguities.push((index_a, index_b));
```
Comment (Ratysz) [inline crates/bevy_ecs/src/schedule/stage.rs:514]: Why these changes? A missing rebase?

File: crates/bevy_ecs/src/schedule/label.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ratysz) [reply crates/bevy_ecs/src/schedule/label.rs:11]: Thinking about this more, do we need to essentially cache the name at this level? As far as I can tell, this is only used for error reporting right now - all of which crash, because they cannot be recovered from. In the future this will be used by the editor, which will most likely cache its own strings, including those with labels in them. I think a `Debug` (and potentially Display) requirement on `Label<T>` is the right way to go.

**Review by Ratysz — APPROVED** [review]

**Review by Ratysz — APPROVED** [review]

File: crates/bevy_app/src/lib.rs [hunk 8]
```diff
@@ -1,7 +1,35 @@
+use bevy_ecs::StageLabel;
+
 /// The names of the default App stages
-pub mod stage;
+#[derive(Debug, Hash, PartialEq, Eq, Clone, StageLabel)]
+pub enum CoreStage {
```
Comment (alice-i-cecile) [inline crates/bevy_app/src/lib.rs:5]: IMO `CoreStage` and `StartupStage` should be marked #[non_exhaustive] to signal that they're not fixed or exhaustive, and avoid people trying to exhaustively pattern match to them.

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_app/src/lib.rs (hunk 8 above)
Comment (TheRawMeatball) [reply crates/bevy_app/src/lib.rs:5]: Why'd they try to match this enum? Values of these types are only produced by user code.

Comment (alice-i-cecile) [reply crates/bevy_app/src/lib.rs:5]: Hmm, that's a fair point. I'm trying to think of cases where you might possibly try but can't.

File: examples/ecs/ecs_guide.rs [hunk 9]
```diff
@@ -309,11 +309,22 @@ fn main() {
...
+        // First, we label the system we want to refer to using `.label`
+        // Then, we use either `.before` or `.after` to describe the order we want the relationship
+        .add_system_to_stage(
+            MyStages::AfterRound,
+            score_check_system.system().label("score_check"),
+        )
+        .add_system_to_stage(
+            MyStages::AfterRound,
+            game_over_system.system().after("score_check"),
+        )
```
Comment (Ratysz) [inline examples/ecs/ecs_guide.rs]: Should probably be a fancier label.

File: crates/bevy_transform/src/lib.rs [hunk 10]
```diff
@@ -23,15 +22,27 @@ impl Plugin for TransformPlugin {
...
+            .add_startup_system_to_stage(
+                StartupStage::PostStartup,
+                transform_propagate_system::transform_propagate_system
+                    .system()
+                    .label("transform_propagate_system")
+                    .after("parent_update_system"),
+            )
+            .add_system_to_stage(
+                CoreStage::PostUpdate,
+                parent_update_system.system().label("parent_update_system"),
```
Comment (Ratysz) [inline crates/bevy_transform/src/lib.rs]: Should definitely be a fancier label.

File: crates/bevy_transform/src/lib.rs [hunk 11]
```diff
@@ -23,15 +22,27 @@ impl Plugin for TransformPlugin {
...
             )
-            .add_system_to_stage(CoreStage::PostUpdate, parent_update_system.system())
             .add_system_to_stage(
                 CoreStage::PostUpdate,
-                transform_propagate_system::transform_propagate_system.system(),
+                transform_propagate_system::transform_propagate_system
+                    .system()
+                    .label("transform_propagate_system")
+                    .after("parent_update_system"),
```
Comment (Ratysz) [inline crates/bevy_transform/src/lib.rs]: Ditto.

**Review by Ratysz — APPROVED** [review]

**Review by alice-i-cecile — APPROVED** [review]

**Comment by Ratysz** [comment]
I've started experimenting on this branch, and some things keep nagging at me.

The derive macro is `#[derive(ThingLabel)]`, but `ThingLabel` is a type, not a trait - this is not how derive is normally used. It's bound to confuse someone.

All methods using labels take `impl Into<ThingLabel>`. This can be minsconstrued as requiring to implement `Into<ThingLabel>` in order to use something as a label for a thing - which is what actually happens in the end, albeit indirectly. Deriving a `ThingLabel` is, again, not the first thing that would come to mind.

In my original prototype `ThingLabel` is a trait, which makes the derive macro a natural fit and allows the arguments to be `impl ThingLabel` (and `&impl ThingLabel`, which could probably be `&dyn ThingLabel`) instead. I'll see what it takes to adapt this branch to that approach, just so we have something to compare against.

File: crates/bevy_ui/src/lib.rs [hunk 12]
```diff
@@ -20,15 +20,21 @@ pub mod prelude {
...
+};
 use bevy_render::render_graph::RenderGraph;
 use update::ui_z_system;
 
 #[derive(Default)]
 pub struct UiPlugin;
-
-pub mod stage {
-    pub const UI: &str = "ui";
+#[derive(Debug, Hash, PartialEq, Eq, Clone, StageLabel)]
```
Comment (cart) [inline crates/bevy_ui/src/lib.rs:31]: In general enum type names should be singular (and i personally think that applies to stage labels);

File: examples/ecs/ecs_guide.rs [hunk 13]
```diff
@@ -239,6 +239,17 @@ fn local_state_system(mut state: Local<State>, query: Query<(&Player, &Score)>)
     state.counter += 1;
 }
 
+#[derive(Debug, Hash, PartialEq, Eq, Clone, StageLabel)]
+enum MyStages {
```
Comment (cart) [inline examples/ecs/ecs_guide.rs:243]: As mentioned in another comment, I think we should be consisent with stage enum naming. Imo this should be `MyStage` for parity with `CoreStage`

**Review by cart — COMMENTED** [review]
This looks good to me, but I agree with @Ratysz's assessment of the trait/derive. 

We should also discuss removing the `Eq` constraint from labels (and changing DynEq to DynPartialEq). Its not _strictly_ needed and would cut down on label declaration boilerplate. However (1) that might not work with @Ratysz 's proposed changes and (2) `Eq` is still correct "semantically", which imo is the main argument to include it. 

I'm also about to push some (hopefully) non-contentious code style changes.

**Comment by TheRawMeatball** [comment]
These labels are used in hashmaps which need Eq, so I don't think removing that is an option. 

For the derives, I initally had an `IntoLabel` trait, and had its derive ask for a custom `#[label_type]` attribute, but @Ratysz changed my mind, so I'm not going to act on that yet.

**Comment by Ratysz** [comment]
> I'll see what it takes to adapt this branch to that approach, just so we have something to compare against.

Here are the results: [diff](https://github.com/bevyengine/bevy/compare/master...Ratysz:labels_as_traits) with Bevy master, [diff](https://github.com/TheRawMeatball/bevy/compare/zst-labels...Ratysz:labels_as_traits) with this branch. I'm biased, but I think this is cleaner.

Things that can be tweaked:
1. Label traits' guts are implemented via a plain macro. This is not as useful right now, but in the future we will likely have more kinds of different labels. The traits' definition itself can also be moved into the macro, but I opted to leave it out for ease of refactoring.
2. `DynClone` functionality is incorporated into the label traits themselves; this creates minor code duplication - although, seemingly less than when trying to lift it into a separate trait. I've opted to `#[doc(hidden)]` the `dyn_clone()`.
3. `type BoxedThingLabel = Box<dyn ThingLabel>` is used internally everywhere (I think). It is not exported, because it bloats the API surface and is trivial.
4. Some `impl ThingLabel` arguments could be replaced with `&dyn ThingLabel`, like [here](https://github.com/Ratysz/bevy/blob/labels_as_traits/crates/bevy_ecs/src/schedule/mod.rs#L167-L177) (this one is done in order to get rid of an extra generic whenever these methods are used). I've opted to not do that yet, because it adds `&`s to user code; this would be in line with how `std` keys are used, however: take when inserting, borrow when querying.

File: crates/bevy_ecs/macros/src/lib.rs [hunk 14]
```diff
@@ -466,3 +466,35 @@ pub fn derive_system_param(input: TokenStream) -> TokenStream {
...
+#[proc_macro_derive(StageLabel)]
+pub fn derive_stage_label(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+    derive_label(input, Ident::new("StageLabelMarker", Span::call_site())).into()
+}
+
+fn derive_label(input: DeriveInput, label_type: Ident) -> TokenStream2 {
+    let ident = input.ident;
+
+    let manifest = Manifest::new().unwrap();
```
Comment (jakobhellermann) [inline crates/bevy_ecs/macros/src/lib.rs:485]: This won't work for examples and doc tests if `bevy_ecs` is used in the main dependencies and `bevy` in the `dev-dependencies`.
Searching in the main deps first and otherwise in the dev-dependencies like [here](https://github.com/bevyengine/bevy/pull/1426/files#diff-5ef94ef273f2d90770db3f44049b344e067a2e36349fba82112dd389b2ad803fR453) should work better.

Comment (Ratysz) [reply crates/bevy_ecs/macros/src/lib.rs:485]: [This test](https://github.com/TheRawMeatball/bevy/blob/zst-labels/crates/bevy_ecs/src/schedule/system_descriptor.rs#L19-L31) passes.

Comment (jakobhellermann) [reply crates/bevy_ecs/macros/src/lib.rs:485]: If `bevy` was added to the dev-dependencies it would fail. Or rather, the test would work but the other derives would fail.

Comment (cart) [reply crates/bevy_ecs/macros/src/lib.rs:485]: yeah its probably worth adopting that method here.

**Comment by cart** [comment]
@TheRawMeatball 
> These labels are used in hashmaps which need Eq, so I don't think removing that is an option.

But the user-defined label type isn't actually the type used in the hashmap (in the original impl we were discussing). We could remove the Eq impl in the user-defined type and still implement Eq on the wrapper type that is actually stored. I know this works because I implemented it before leaving the comment  :smile: 

@Ratysz 
I'm also partial to this impl. Definitely simpler in general. I also do like `get_stage(label: &dyn StageLabel)`, not because of parity with std (afaik refs are only used in those interfaces because they work best generically. imo copy types shouldn't use refs in general). I like it because it makes calling the relevant methods less weird when it comes to generic type parameters.

**Comment by Ratysz** [comment]
So... What should we do then? `&dyn ThingLabel` where appropriate then pull my branch into this one?

**Comment by cart** [comment]
Hmm I'm guessing you don't have the ability to push commits here. I think the only important outcome is both authors getting credited. So if you create a new pr with @TheRawMeatball's commits in the history, then that works. Or alternatively @TheRawMeatball can apply your commits to this pr.

**Comment by cart** [comment]
And to be clear, i think we should only use `&dyn ThingLabel` when it makes the "calling ergonomics" situation more natural. We should accept `impl ThingLabel` wherever possible imo.

**Comment by YohDeadfall** [comment]
> Or alternatively @TheRawMeatball can apply your commits to this pr.

It's also possible to open a PR to his repo and merge changes with rebase, so no merge commit will be in the history of this PR.

**Comment by Ratysz** [comment]
Something went very, very wrong during conflict resolution. The PR lives on as #1473.

**Comment by Ratysz** [comment]
1473 is now merged, closing.
